package-builder/flox-build.mk

```diff
@@ -378,49 +357,86 @@ define BUILD_local_template =
   # directory and FLOX_ENV.
   $(eval $(_pvarname)_out = /tmp/store_$($(_pvarname)_hash)-$(_name))
 
-  # Our aim in performing a manifest build is to replicate as closely as possible
-  # the experience of running those same build script commands from within an
-  # interactive `flox activate -m dev` shell (i.e. using the "develop" environment).
-
-  # But unlike the interactive case, the manifest build seeks to use dependencies
-  # as found in the target package's "wrapper" environment in preference to those
-  # found in the "develop" environment. It does this for the express purpose of
-  # preventing the resulting closure from depending on the "develop" environment,
-  # which can contain compilers, libraries and tools not required at runtime.
-
-  # The way it does this is by performing the build within a nested activation of
-  # each of the "develop" and "wrapper" environments:
-  # * the [outer] "develop" environment is activated first, providing access to
-  #    runtime dependencies and compilers/tools required only at build time
-  # * then the [inner] "wrapper" environment is activated, providing access
-  #    to only those runtime dependencies
-
-  # The only issue with the above approach is that references to the "develop"
-  # environment can leak into the resulting build. For example, each of these
-  # activations can prepend to a PYTHONPATH that gets embedded in the
-  # build, which has the effect of pulling both environments into the closure.
-
-  # We could use `env -i` to prevent all variables from leaking from the
-  # "develop" environment into the inner activation, but then that causes
-  # problems for compilers that rely on NIX_CC* environment variables set in
-  # the outer activation. To address this problem we use the `env-filter` script
-  # to remove all variables not specifically allowed through to the inner
-  # activation.
-
-  # The final result is approximately the following:
-  #   $(FLOX_INTERPRETER)/activate ... -- \
-  #     $(_libexec_dir)/env-filter $(env_filter_ALLOW_ARGS) -- \
-  #       $(_build_wrapper_env)/wrapper ... -- bash -e buildScript
+  # Our aim in performing a manifest build is to replicate as closely as
+  # possible the experience of running those same build script commands
+  # from within an interactive `flox activate -m dev` shell (i.e. using
+  # the "develop" environment), but we also need to provide a way to avoid
+  # having to materialize compilers and tools not required at runtime, and
+  # our approach for addressing this issue is the following:
+
+  # 1. create "build-wrapper" environments for each build, customized to only
+  #    include packages as found in the "runtime-packages" manifest attribute
+  # 2. perform the build using the "develop" environment
+  # 3. replace all references to the "develop" environment path with that of
+  #    the "build-wrapper" environment
```

Comment:
AFIACT this does not prevent inclusion of paths outside of the "build-wrapper" env i.e. the wider "develop" env.
Any tool embedding `realpath`ed paths will potentially expand its closure outside of the build-wrapper.

Reply:
Correct - our approach is to _audit_ not _prevent_ the inclusion of disallowed paths. This gives us a couple of advantages:
* a build that is successful from a Nix perspective can be inspected with `nix why-depends` to know why something was brought in by accident - by comparison a failed build doesn't exist in the database to be inspected at all (even if it's present in the store)
* we can implement out-of-band logic to warn of unexpected packages in the closure, which is helpful when helping the user to step up the purity of their builds in increments